backend/src/main/java/ai/verta/modeldb/metadata/MetadataDAORdbImpl.java

```diff
@@ -83,15 +93,83 @@ public boolean addLabels(IdentificationType id, List<String> labels) {
     }
   }
 
+  private void processLabels(
+      Session session,
+      IdentificationType id,
+      List<String> labels,
+      Function4<Supplier<Object>, Object, String, Object> processLabel) {
+    if (id.getIdType() == VERSIONING_REPO_COMMIT_BLOB_DESCRIPTION) {
+      String label = labels.get(0);
+      LabelsMappingDescriptionEntity.LabelMappingId id0 = LabelsMappingDescriptionEntity.createId(id);
+      LabelsMappingEntity existingLabelsMappingEntity =
+              session.get(LabelsMappingEntity.class, id0);
+      processLabel.apply(() -> new LabelsMappingDescriptionEntity(id0, label), existingLabelsMappingEntity, label);
+    } else {
+      for (String label : labels) {
+        LabelsMappingEntity.LabelMappingId id0 = LabelsMappingEntity.createId(id, label);
+        LabelsMappingEntity existingLabelsMappingEntity =
+            session.get(LabelsMappingEntity.class, id0);
+        processLabel.apply(() -> new LabelsMappingEntity(id0), existingLabelsMappingEntity, label);
+      }
+    }
+  }
+
+  private <T> void saveLabel(
+      Session session, Supplier<T> labelsMappingEntity, T existingLabelsMappingEntity, String label) {
+    if (existingLabelsMappingEntity == null) {
+      session.save(labelsMappingEntity.get());
+    } else {
+      Status status =
+          Status.newBuilder()
+              .setCode(Code.ALREADY_EXISTS_VALUE)
+              .setMessage("Label '" + label + "' already exists with given ID")
+              .build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+  }
+
+  private <T> void deleteLabels(
+      Session session, Supplier<T> labelsMappingEntity, T existingLabelsMappingEntity, String label) {
+    if (existingLabelsMappingEntity != null) {
+      session.delete(existingLabelsMappingEntity);
+    } else {
+      Status status =
+          Status.newBuilder()
+              .setCode(Code.NOT_FOUND_VALUE)
+              .setMessage("Label '" + label + "' not found in DB")
+              .build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+  }
+
   @Override
   public List<String> getLabels(IdentificationType id) {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
-      Query query = session.createQuery(GET_LABELS_HQL);
-      query.setParameter("entityHash", getEntityHash(id));
-      query.setParameter("entityType", id.getIdTypeValue());
-      List<LabelsMappingEntity> labelsMappingEntities = query.list();
+      List<LabelsMappingEntityBase> labelsMappingEntities;
+      if (id.getIdType() == VERSIONING_REPO_COMMIT_BLOB_DESCRIPTION) {
```

Comment:
this if should be on what metadata is being processed , either label or description.

Reply:
this function could change to List<String> getMetdata(IdentificationType id, MetadataType mt) and the if can be 
```
if(mt = MetadataType.Label){
...}
else if(mt = MetadataType.Description){
...
}

```

MetadataType could be internal enum

Reply:
now it's in separate methods